folly/CPortability.h

```diff
@@ -53,4 +53,725 @@
+#define S_IRGRP S_IRUSR
+#define S_IXOTH S_IXUSR
+#define S_IWOTH S_IWUSR
+#define S_IROTH S_IRUSR
+#endif
+
+
+#ifdef _MSC_VER
+
+#ifndef __STDC__
+#define __STDC__ 1
+#include <io.h>
+#undef __STDC__
+#else
+#include <io.h>
+#endif
+
+#include <assert.h>
+#include <direct.h>
+#include <intrin.h>
+#include <malloc.h>
+#include <math.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <Windows.h>
+#include <WinSock2.h>
+
+#include <pthread.h>
+#include <pthread_clock.h>
+
+
+// Do nothing for the system log for now.
+inline void openlog(const char*, int, int) {}
+inline void closelog() {}
+inline void syslog(int, const char*, ...) {}
+
+#define SIGCHLD 0
+// We use SIGILL as SIGALRM, as SIGILL is
+// never generated by windows.
+#define SIGALRM SIGILL
+// All 4 of these are invalid signal IDs.
+#define SIGVTALRM 26
+#define SIGPROF 27
+#define SIGKILL 28
+#define SIGUSR1 29
+
+inline void __stdcall __alarm_callback_func(HWND, UINT, UINT_PTR, DWORD) {
+  raise(SIGALRM);
+}
+
+inline unsigned int alarm(unsigned int seconds) {
+  SetTimer(NULL, 0, (UINT)(seconds * 1000), __alarm_callback_func);
+  return 0;
+}
+
+inline char* asctime_r(const tm* tm, char* buf) {
+  char tmpBuf[64];
+  if (asctime_s(tmpBuf, sizeof(tmpBuf), tm))
+    return NULL;
+  return strcpy(buf, tmpBuf);
+}
+
+inline void bzero(void* s, size_t n) {
+  ZeroMemory(s, n);
+}
+
+#define DT_UNKNOWN 0
+#define DT_DIR     1
+#define DT_REG     2
+#define DT_LNK     3
+struct dirent {
+  unsigned char d_type;
+  char* d_name;
+};
+
+struct DIR {
+  dirent dir;
+  HANDLE hand;
+  int cnt;
+  char name[MAX_PATH * 3];
+};
+
+inline int closedir(DIR* dir) {
+  FindClose(dir->hand);
+  free(dir);
+  return 0;
+}
+
+inline char* ctime_r(const time_t* t, char* buf) {
+  char tmpBuf[64];
+  if (ctime_s(tmpBuf, sizeof(tmpBuf), t))
+    return NULL;
+  return strcpy(buf, tmpBuf);
+}
+
+inline int dprintf(int fd, const char *fmt, ...) {
+  va_list args;
+  va_start(args, fmt);
+
+  int len = vsnprintf(NULL, 0, fmt, args);
+  if (len <= 0)
+    return -1;
+  char* buf = (char*)malloc(len + 1);
+  if (vsnprintf(buf, len + 1, fmt, args) == len &&
+      _write(fd, buf, len) == len) {
+    free(buf);
+    va_end(args);
+    return len;
+  }
+  
+  free(buf);
+  va_end(args);
+  return -1;
+}
+
+inline int finite(double d) {
+  return isfinite(d) ? 1 : 0;
+}
+
+inline int getgid() {
+  return 0;
+}
+
+#ifndef _PID_T_
+typedef uintptr_t pid_t;
+#define _PID_T_     1
+#endif
+// No major need to implement this, and getting a non-potentially
+// stale ID on windows is a bit involved.
+inline pid_t getppid() {
+  return 0;
+}
+
+typedef size_t rlim_t;
+struct rlimit {
+  rlim_t rlim_cur;
+  rlim_t rlim_max;
+};
+#define RLIMIT_CORE 0
+#define RLIMIT_NOFILE 0
+#define RLIMIT_DATA 0
+#define	RLIMIT_STACK 3
+#define RLIM_INFINITY SIZE_MAX
+
+inline int getrlimit(int type, rlimit* dst) {
```

Comment:
Mark the function as `_Success_( return == 0 )` and `dst` as `_Out_` to detect uninit memory usage.